### PR TITLE
Version 1 of tweetme deployed

### PR DESCRIPTION
**Pull request by realslimshanky**
Version 1 includes Token/Key management using config.json
Tweets.txt which includes posts in everyline
Unit Testing of entire bot using unittest module

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,47 @@
...
+- Create a file named `tweets.txt` whose each line should be in below format
+
+`What are dotfiles? :: http://thegeekyway.com/what-are-dotfiles/`
+`[Ultimate Guide] How SSH works? :: http://thegeekyway.com/ultimate-guide-how-ssh-works/`
+
+- As you can see, each line has two parts separated by `::`. It's upto you what the content should be for each part.
+
+### The First Run
+
+- Run the command `pipenv run python tweetme.py`. You'll get a message asking you to add Twitter API configuration inside config.txt. Use [Twitter Application Management](https://apps.twitter.com) and create a new application.
```
Comment (CuriousLearner) [inline README.md]: config.txt can be put in back quotes here.

File: testtweetme.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import tweetme
+from birdy.twitter import UserClient
+
+
+class TestConfigFile(unittest.TestCase):
+    """
+    This test module will run only to test manage_token() method.
+    """
+
+    def test_1_nofile(self):
```
Comment (CuriousLearner) [inline testtweetme.py]: No need to define tests having numbers. The name of the tests should be self-explanatory.

File: testtweetme.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        """
+
+        try:
+            os.remove("testconfig.json")
+            self.assertTrue(True)
+        except:
+            self.assertTrue(False)
+
+
+@unittest.skipIf(not tweetme.manage_token(), ("Twitter API cannot be tested if config.json is "
```
Comment (CuriousLearner) [inline testtweetme.py]: Can we fix indentation in the decorator please?

File: tweetme.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        with open(turn_file, mode='w') as f:
+            f.write('1')
+        with open(tweets_file, mode='r') as f:
+            Tweet =  f.readline()
+        return Tweet.split("::")
+    else:
+        """When both files are present, check turn.txt and use it's value as index to
+        find the next tweet from tweets.txt and write index + 1 in turn.txt"""
+        with open(turn_file, mode='r') as f:
+            turn = int(f.readline())
```
Comment (CuriousLearner) [inline tweetme.py]: `turn` variable doesn't really explain the purpose, can we name this something else?

File: tweetme.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        return tweets[turn].split("::")
+
+
+def manage_twitter_client():
+    """
+    This function will create twitter client using configurations and send tweet.
+    """
+
+    configError = ("Please open config.json file located in the project "
+        "directory and relace the value '0' of all the tokens and keys in "
```
Comment (CuriousLearner) [inline tweetme.py]: small typo `relace` -> `replace`

File: tweetme.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    keys = manage_token()
+    if not keys:
+        print(configError)
+    else:
+        tweet = get_tweet()
+        if tweet:
+            client = UserClient(*keys)
+            response = client.api.statuses.update.post(status='{} \n {}'.format(tweet[0], tweet[1]))
+            print('You tweet is out in the world.\nCheck it out https://twitter.com/{}/status/{}'.format(
```
Comment (CuriousLearner) [inline tweetme.py]: Can we break this line to be more readable?

Also, we can log the data instead of just printing on the console.

File: tweetme.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            f.write('0')
+        return False
+    
+    elif turn_file not in os.listdir():
+        """When turn.txt is not present, creates a new turn.txt and writes 1 in it
+        and return the first tweet from tweets.txt"""
+        with open(turn_file, mode='w') as f:
+            f.write('1')
+        with open(tweets_file, mode='r') as f:
+            Tweet =  f.readline()
```
Comment (CuriousLearner) [inline tweetme.py]: variable `Tweet` can be lowercase. Also, it seems better that we use `tweet_text` instead.

**Review by CuriousLearner — CHANGES_REQUESTED** [review]
Overall looks good. I've just few minor comments that you can look into.

Also, instead of printing on console, we can use logging to keep track of this both on the console and the file.

File: testtweetme.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    
+    def test_4_removing_config(self):
+        """
+        This test only removes config.json created and used by above tests.
+        """
+
+        try:
+            os.remove("testconfig.json")
+            self.assertTrue(True)
```
Comment (CuriousLearner) [inline testtweetme.py]: What's the use of asserting True here?

What are we trying to accomplish here?

File: testtweetme.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                             "not present with valid Token/Keys."))
+class TestTwitterAPI(unittest.TestCase):
+    """
+    This module is used to test Twitter API with the given configurations.
+    """
+
+    username = 'TheGeekyWay'
+    keys = tweetme.manage_token()
+
+    def test_1_api_call(self):
```
Comment (CuriousLearner) [inline testtweetme.py]: Ideally, tests should not depend on network calls.

But I think for testing creds, this is okay.

File: testtweetme.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
+import os
```
Comment (CuriousLearner) [inline testtweetme.py]: Can we make a test directory and place tests separately in different files?

**Comment by CuriousLearner** [comment]
Since the tests are there (would need refactoring) I think we can also include `.travis.yml` to run the tests.

Also, after again thinking on testing the credential thing, I think we can ignore that. Doesn't make sense.

File: testtweetme.py (hunk 2 above)
Comment (realslimshanky) [reply testtweetme.py]: The problem was I needed the tests to run in a specific order and unittest runs the test in the alphabetical order of naming. Hence, I had to include the numbers. I'll add a note about this too in order to self document about the importance of numbers.

File: testtweetme.py (hunk 3 above)
Comment (realslimshanky) [reply testtweetme.py]: I think it's where it is supposed to be. As it's been applied to entire TestTwitterAPI class.

File: tweetme.py (hunk 4 above)
Comment (realslimshanky) [reply tweetme.py]: What about `next_tweet_index`?

File: tweetme.py (hunk 6 above)
Comment (realslimshanky) [reply tweetme.py]: Sure, I'll add logging module to help us out.

File: tweetme.py (hunk 7 above)
Comment (realslimshanky) [reply tweetme.py]: Yeah, that seems more like it.

File: testtweetme.py (hunk 8 above)
Comment (realslimshanky) [reply testtweetme.py]: At the end of the previous tests, we have to remove the `testconfig.json` file. If the operation is successful user can be notified of the test getting passed.

File: testtweetme.py (hunk 10 above)
Comment (realslimshanky) [reply testtweetme.py]: Sure. let me take care of it.

**Comment by realslimshanky** [comment]
Would be better if we create another issue/PR for `.travis.yml` as that might require separate discussion.

File: tests/test_config.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+class TestConfigFile(unittest.TestCase):
+    """
+    This test module will run only to test manage_token() method.
+
+    Note: Numbers just after `test_` are important as they facilitate test to run in
+          specific order as `unittest` runs test in alphabetical order of thier names.
+    """
+
+
+    def test_1_nofile(self):
```
Comment (CuriousLearner) [inline tests/test_config.py]: No need of numbers in the test name.

File: tests/test_config.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        """
+        Test the module when no config.json is present in the project directory.
+        """
+
+        self.assertFalse(tweetme.manage_token('testconfig.json'))
+
+    
+    def test_2_initialfile(self):
+        """
+        Test the module when config.json is present in the project directory.
```
Comment (CuriousLearner) [inline tests/test_config.py]: The doc-strings are not required to start with `Test that` or something similar. All tests are testing something :)

File: tests/test_config.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        """
+
+        try:
+            os.remove("testconfig.json")
+            self.assertTrue(True)
+        except:
+            self.assertTrue(False)
+
+
+'''
```
Comment (CuriousLearner) [inline tests/test_config.py]: Everything below this line is redundant. It is repeating in every file. Can we do better?

File: tests/test_config.py (hunk 11 above)
Comment (realslimshanky) [reply tests/test_config.py]: I'll go with an alternate naming.

File: tests/test_config.py (hunk 12 above)
Comment (realslimshanky) [reply tests/test_config.py]: Okay, let me rephrase that.

File: tests/test_config.py (hunk 13 above)
Comment (realslimshanky) [reply tests/test_config.py]: It allows the tests to run individually. A better option would be placing different tests in just one file which we previously discarded.

File: .gitignore [hunk 14]
```diff
@@ -99,3 +99,10 @@ ENV/
 
 # mypy
 .mypy_cache/
+
+# config/tweet files
+config.json
```
Comment (CuriousLearner) [inline .gitignore:104]: We can add a `sample.config.json` listing all the variables needed to run this app.

Users can then simply rename this file to `config.json` with appropriate values.

File: .gitignore [hunk 15]
```diff
@@ -99,3 +99,10 @@ ENV/
...
+turn.txt
+tweets.txt
+next_tweet_index.txt
+tweetme.log
```
Comment (CuriousLearner) [inline .gitignore:108]: Ideally it should be generated somewhere under `/etc/` but I think it is okay for now.

File: tweetme.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
+import os
+import json
+import twitter
+import logging
+from birdy.twitter import UserClient
+
+logging.basicConfig(filename='tweetme.log', format='%(asctime)s %(message)s', level=logging.DEBUG)
+
+def manage_token(filename="config.json"):
```
Comment (CuriousLearner) [inline tweetme.py]: a better name could be `get_config_from_file`

File: tweetme.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+                return (config["consumer_key"],
+                        config["consumer_secret"],
+                        config["access_token"],
+                        config["access_token_secret"],
+                )
+            else:
+                return False
+
+
+def get_tweet(tweets_file='tweets.txt', turn_file='next_tweet_index.txt'):
```
Comment (CuriousLearner) [inline tweetme.py]: rename this to `get_next_tweet_from_file`

File: tests/path.py [hunk 18]
```diff
@@ -0,0 +1,14 @@
+import os
+import sys
+
+'''
+Since tweetme.py is now present in parent directory we need to add appropriate path in order to
```
Comment (CuriousLearner) [inline tests/path.py:5]: Not sure why this is needed. I still need to test this code locally once. But till the time you can look into other things.

**Review by CuriousLearner — CHANGES_REQUESTED** [review]
Just nit-picks :)

File: .gitignore (hunk 14 above)
Comment (realslimshanky) [reply .gitignore:104]: Right now the `config.json` is created by the program itself with all the appropriate fields on the first run and user is notified of the changes they have to make in order to make it work.

File: .gitignore (hunk 15 above)
Comment (realslimshanky) [reply .gitignore:108]: We can shift them if there are enough files for us to handle. Right now the repository looks clean since all these files are ignored by default.

File: tests/path.py (hunk 18 above)
Comment (realslimshanky) [reply tests/path.py:5]: In order to `import tweetme` which is not in the parent directory where all the tests files are, we had to add the location of the folder where `tweetme` is in order to access it as a module. This is where `path.py` comes it.

File: .gitignore (hunk 14 above)
Comment (CuriousLearner) [reply .gitignore:104]: Yeah, I remember that. Thanks for bringing it up again. I forgot about this in the last review pass.

**Comment by CuriousLearner** [comment]
Will need to put in pytest and refactor this. But looks okay for now.

**Review by CuriousLearner — APPROVED** [review]
